technical-reference/reference-technical-articles/diamond-comptroller.md

```diff
@@ -0,0 +1,83 @@
+- Comptroller implementation before enabling the Diamond Proxy:  https://bscscan.com/address/0x909dd16b24cef96c7be13065a9a0eaf8a126ffa5
+
+The following diagram shows the previous situation:
+
+<figure><img src="../../.gitbook/assets/diamond-previous.png" alt="Diamond - Previous hierarchy of contracts"><figcaption></figcaption></figure>
+
+The Unitroller contract delegated calls to the Comptroller. The Comptroller had the storage layout defined by ComptrollerV12Storage, which extended the UnitrollerAdminStorage.
+
+The Diamond Proxy pattern is widely used in Solidity development for its ability to separate concerns and improve code maintainability. It is often used in complex contracts that require multiple facets with shared functionality.
+
+## EIP - 2535 Overview
+1. The Diamond Proxy pattern involves creating a "proxy" contract that acts as a single entry point for all functionality shared by multiple contracts. This allows contracts to share functionality without having to duplicate code in each contract.
+
+2. Each contract that needs to share functionality with the proxy contract is referred to as a "facet." Facets are separate contracts that can be upgraded or modified independently of each other and the proxy contract.
+
+3. The proxy contract delegates function calls to the correct facet contract based on the function selector. This allows the proxy contract to act as a single entry point for all shared functionality.
+
+4. The Diamond Proxy pattern also supports multiple inheritances. This means that each facet can inherit from other contracts, including other facets. This allows for a high degree of code reuse and modular design.
+
+## Enabling Venus Integration with the existing Unitroller - Diamond and Transparent proxy combination
+
+**Step 1:** The first step involved the creation of a new contract that serves as the Diamond proxy. This contract contains a mapping to hold the addresses of the facet contracts and a fallback function that delegates the function call to the appropriate facet. 
+
+Chained delegate calls are used, meaning calldata is delegated from Unitroller to the Diamond contract, and then a chained delegate call is made to a specific facet based on the function selector. 
+
+<figure><img src="../../.gitbook/assets/diamond-facets.png" alt="Diamond facets"><figcaption></figcaption></figure>
+
+The Diamond contract checks for the facet address in the **selectors-to-facet-address** mapping (`_selectorToFacetAndPosition` internal variable) and then makes a delegate call to that facet address. Facets inherit the ComptrollerStorage (ComptrollerV13Storage) to access the states.
+
+**Step 2:** Division of the comptroller into multiple facets based on the separation of concerns.
+
+Following the same eternal storage pattern i.e. inheriting storage contract for storage slots. This basically replaced the AppStorage of Diamond pattern.
```

Comment:
Generally [AppStorage](https://eips.ethereum.org/EIPS/eip-2535#storage) is used with the diamond pattern. Since the Core Pool contracts still use Solidity 0.5.16 this wasn't an option and storage is handled using the same pattern as the original comptroller, by inheriting storage contracts.

Reply:
Done [aa851b5](https://github.com/VenusProtocol/venus-protocol-documentation/pull/69/commits/aa851b5af5063542068ad5e6bf0ab9fb41087b00)